### PR TITLE
Loop unrolling

### PR DESCRIPTION
**Pull request by sjalander**

### CODE REVIEW

File: tests/libjlm/opt/test-unroll.cpp [hunk 1]
```diff
@@ -5,6 +5,7 @@
 
 #include "test-operation.hpp"
 #include "test-registry.hpp"
+#include "test-types.hpp"
```
Comment (phate) [inline tests/libjlm/opt/test-unroll.cpp:8]: Do you actually need this include? I do not see that you defined any of the types in the new test case.

File: tests/libjlm/opt/test-unroll.cpp [hunk 2]
```diff
@@ -157,9 +158,9 @@ test_known_boundaries()
 		auto end = jive::create_bitconstant(graph.root(), 32, 4);
 
 		auto theta = create_theta(ult, add, init, step, end);
-//		jive::view(graph, stdout);
```
Comment (phate) [inline tests/libjlm/opt/test-unroll.cpp:160]: This is just noise, please add the comments before the jive::view calls again.

File: tests/libjlm/opt/test-unroll.cpp [hunk 3]
```diff
@@ -270,11 +271,95 @@ test_unknown_boundaries()
...
+	auto inner2_compare = jive::bitult_op::create(32, inner2_add, lvi2_end->argument());
+	auto inner2_match = jive::match(1, {{1, 1}}, 0, 2, inner2_compare);
+	inner2_theta->set_predicate(inner2_match);
+	lvi2_init->result()->divert_to(inner2_add);
+
+	
+	jive::view(graph, stdout);
+	jlm::loopunroll loopunroll(4);
+	loopunroll.run(rm, sd);
+	/* FIX ME! Need a way to check that only the inner loop gets unrolled */
```
Comment (phate) [inline tests/libjlm/opt/test-unroll.cpp:346]: Can we please add these inserts (and remove the FIXMEs)?

File: libjlm/include/jlm/opt/unroll.hpp [hunk 4]
```diff
@@ -34,6 +34,14 @@ class loopunroll final : public optimization {
 	: factor_(factor)
 	{}
 
+	/**
+	* Given a module all inner most loops (thetas) are found and unrolled if possible.
+        * All nodes in the module are traversed and if a theta is found and is the inner most theta
+        * then an attempt is made to unroll it.
+	*
+	* \param module The module to attempt loop unrolling on.
```
Comment (phate) [inline libjlm/include/jlm/opt/unroll.hpp:42]: \param module Module where the innermost loops are unrolled

File: libjlm/include/jlm/opt/unroll.hpp [hunk 5]
```diff
@@ -225,6 +233,13 @@ class unrollinfo final {
 	jive::argument * idv_;
 };
 
+
+/**
+* Try to unroll the given theta.
+*
+* \param theta The theta to attempt the unrolling on.
+* \param factor The number of times to unroll the loop.
```
Comment (phate) [inline libjlm/include/jlm/opt/unroll.hpp:247]: Can we add an example here to specify what factor means? Like if the factor is 2, then the loop body is duplicated in the unrolled loop (if that is actually true).

File: libjlm/src/opt/unroll.cpp [hunk 6]
```diff
@@ -484,51 +458,47 @@ unroll(jive::theta_node * otheta, size_t factor)
...
 		if (auto structnode = dynamic_cast<jive::structural_node*>(node)) {
 			for (size_t n = 0; n < structnode->nsubregions(); n++)
-				unroll(structnode->subregion(n), factor);
-
-			if (auto theta = dynamic_cast<jive::theta_node*>(node))
-				unroll(theta, factor);
+				unrolled = unroll(structnode->subregion(n), factor);
+
+			/* Try to unroll if an inner loop hasn't already been found */
+			if (!unrolled)
```
Comment (phate) [inline libjlm/src/opt/unroll.cpp:510]: Can we use {} for this if-statement here?

File: libjlm/src/opt/unroll.cpp [hunk 7]
```diff
@@ -205,105 +205,145 @@ unroll_body(
 	theta->subregion()->copy(target, smap, false, false);
 }
 
+/*
+	Copy the body of the thet theta and unroll it ntimes.
```
Comment (phate) [inline libjlm/src/opt/unroll.cpp:209]: remove thet

Comment (phate) [reply libjlm/src/opt/unroll.cpp:209]: Also, you say here ntimes, but the parameter is called factor.

File: libjlm/src/opt/unroll.cpp [hunk 8]
```diff
@@ -432,7 +472,6 @@ unroll_unknown_theta(const unrollinfo & ui, size_t factor)
 			smap.insert(*olv, xv);
 		}
 	}
-
```
Comment (phate) [inline libjlm/src/opt/unroll.cpp:461]: Noise, please reinsert.

Comment (sjalander) [reply libjlm/src/opt/unroll.cpp:461]: So you like meaningless whitespaces :)

File: tests/libjlm/opt/test-unroll.cpp (hunk 3 above)
Comment (sjalander) [reply tests/libjlm/opt/test-unroll.cpp:350]: Not sure how to check for this.
Possibly by counting the number of nodes in the thetas, but it's not really checking for correctness.
Maybe could do check that the number of nodes is greater than the number of nodes of the original body.

Comment (phate) [reply tests/libjlm/opt/test-unroll.cpp:350]: The checking for (complete) correctness will be messy. I would not advice it. Simply counting the number of nodes is sufficient for now.

File: libjlm/src/opt/unroll.cpp (hunk 8 above)
Comment (phate) [reply libjlm/src/opt/unroll.cpp:461]: Fair enough. Leave it as it is now. Thanks!

**Pull request merged**